include/swift/AST/AvailabilitySpec.h

```diff
@@ -215,6 +215,19 @@ class OtherPlatformAvailabilitySpec : public AvailabilitySpec {
   }
 };
 
+/// Maps of macro name and version to availability specifications.
+/// Organized as two nested \c DenseMap keyed first on the macro name then
+/// the macro version. This structure allows to peek at macro names before
+/// parsing a version tuple.
+class AvailabilityMacroMap {
+public:
+  typedef llvm::DenseMap<llvm::VersionTuple,
+                         SmallVector<AvailabilitySpec *, 4>> VersionEntry;
+
+  bool WasParsed = false;
+  llvm::DenseMap<StringRef, VersionEntry> Impl;
```

Comment:
Should this be a StringMap rather than depend on the StringRefs living forever?

Reply:
I think that's safe currently. The StringRefs should point to a buffer attached to the SourceManager of the ASTContext, so both the cache and the sources should be freed together. Does this sound right?

Reply:
Currently one of these buffers is allocated for each `-define-availability` argument. I suspect this doesn't help with the general performances of the `findBuffer` services on wide uses of availability macros, we could look into deregistering these small buffers after parsing and copy all these strings.